src/lib/homestore_backend/replication_state_machine.cpp

```diff
@@ -238,135 +238,219 @@ std::shared_ptr< homestore::snapshot_context > ReplicationStateMachine::last_sna
-            return -1;
-        }
+    auto obj_id = objId(snp_obj->offset);
+    log_str = fmt::format("{} shard_seq_num={} batch_num={}", log_str, obj_id.shard_seq_num, obj_id.batch_id);
 
-        // Use the shard sequence number provided by the follower and we restart the batch.
-        pg_iter->cur_shard_seq_num_ = shard_seq_num;
-        pg_iter->cur_blob_id_ = -1;
-        pg_iter->cur_snapshot_batch_num = 0;
+    LOGD("read current snp obj {}", log_str)
+    //invalid Id
+    if (!pg_iter->update_cursor(obj_id)) {
+        LOGW("Invalid objId in snapshot read, {}, current shard_seq_num={}, current batch_num={}",
+             log_str, pg_iter->cur_obj_id_.shard_seq_num, pg_iter->cur_obj_id_.batch_id);
+        return -1;
     }
 
-    if (pg_iter->end_of_scan()) {
-        // No more shards to read, baseline resync is finished after this.
-        snp_data->is_last_obj = true;
-        LOGD("Read snapshot reached is_last_obj true {}", log_str);
+    //pg metadata message
+    //shardId starts from 1
+    if (obj_id.shard_seq_num == 0) {
+        if (!pg_iter->create_pg_snapshot_data(snp_obj->blob)) {
+            LOGE("Failed to create pg snapshot data for snapshot read, {}", log_str);
+            return -1;
+        }
         return 0;
     }
 
-    // Get next set of blobs in the batch.
-    std::vector< HSHomeObject::BlobInfoData > blob_data_vec;
-    bool end_of_shard;
-    auto result = pg_iter->get_next_blobs(HS_BACKEND_DYNAMIC_CONFIG(max_num_blobs_in_snapshot_batch),
-                                          HS_BACKEND_DYNAMIC_CONFIG(max_snapshot_batch_size_mb) * 1024 * 1024,
-                                          blob_data_vec, end_of_shard);
-    if (result != 0) {
-        LOGE("Failed to get next blobs in snapshot read result={} {}", result, log_str);
-        return -1;
+    // shard metadata message
+    if (obj_id.batch_id == 0) {
+        if (!pg_iter->generate_shard_blob_list()) {
+            LOGE("Failed to generate shard blob list for snapshot read, {}", log_str);
+            return -1;
+        }
+        if (!pg_iter->create_shard_snapshot_data(snp_obj->blob)) {
+            LOGE("Failed to create shard meta data for snapshot read, {}", log_str);
+            return -1;
+        }
+        return 0;
     }
 
-    // Create snapshot flatbuffer data.
-    pg_iter->create_blobs_snapshot_data(blob_data_vec, snp_data->blob, end_of_shard);
-    if (end_of_shard) {
-        pg_iter->cur_snapshot_batch_num = 0;
-    } else {
-        pg_iter->cur_snapshot_batch_num++;
+    // general blob message
+    if (!pg_iter->create_blobs_snapshot_data(snp_obj->blob)) {
+        LOGE("Failed to create blob batch data for snapshot read, {}", log_str);
+        return -1;
     }
-
-    LOGT("Read snapshot num_blobs={} end_of_shard={} {}", blob_data_vec.size(), end_of_shard, log_str);
     return 0;
 }
 
 void ReplicationStateMachine::write_snapshot_obj(std::shared_ptr< homestore::snapshot_context > context,
```

Comment:
I have a question regarding the handling of a specific scenario during the baseline resync process. Could you please clarify how the system currently manages the situation where a follower crashes and the leader's `snapshot_sync_ctx` does not time out? Understanding this would be very helpful.

Reply:
Follower should persist its transmission checkpoint during the baseline resync, so that it can resume from the point where it stops. This logic is not implemented yet.